File: simple-stack-example-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/addoredittask/AddOrEditTaskPresenter.java
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.addoredittask;

import android.os.Bundle;
import android.support.annotation.Nullable;

import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.Bundleable;
import com.zhuinden.simplestack.HistoryBuilder;
import com.zhuinden.simplestack.StateChange;
import com.zhuinden.simplestackdemoexamplefragments.data.repository.TaskRepository;
import com.zhuinden.simplestackdemoexamplefragments.presentation.objects.Task;
import com.zhuinden.simplestackdemoexamplefragments.presentation.paths.tasks.TasksFragment;
import com.zhuinden.simplestackdemoexamplefragments.presentation.paths.tasks.TasksKey;
import com.zhuinden.simplestackdemoexamplefragments.util.BasePresenter;
import com.zhuinden.simplestackdemoexamplefragments.util.MessageQueue;
import com.zhuinden.simplestackdemoexamplefragments.util.Strings;

import javax.inject.Inject;

import rx.android.schedulers.AndroidSchedulers;

/**
 * Created by Owner on 2017. 01. 27..
 */
// UNSCOPED
public class AddOrEditTaskPresenter
        extends BasePresenter<AddOrEditTaskFragment, AddOrEditTaskPresenter>
        implements Bundleable {
    String title;
    String description;

    @Inject
    TaskRepository taskRepository;

    @Inject
    MessageQueue messageQueue;

    @Inject
    Backstack backstack;

    @Inject
    public AddOrEditTaskPresenter() {
    }

    public void updateTitle(String title) {
        this.title = title;
    }

    public void updateDescription(String description) {
        this.description = description;
    }

    String taskId;

    Task task;

    @Override
    protected void onAttach(AddOrEditTaskFragment coordinator) {
        AddOrEditTaskKey addOrEditTaskKey = coordinator.getKey();
        taskId = addOrEditTaskKey.taskId();
        if(!Strings.isNullOrEmpty(taskId)) {
            taskRepository.findTask(addOrEditTaskKey.taskId()).observeOn(AndroidSchedulers.mainThread()).subscribe(taskOptional -> {
                if(taskOptional.isPresent()) {
                    task = taskOptional.get();
                    if(this.title == null || this.description == null) {
                        this.title = task.title();
                        this.description = task.description();
                        coordinator.setTitle(title);
                        coordinator.setDescription(description);
                    }
                }
            });
        }
    }

    @Override
    protected void onDetach(AddOrEditTaskFragment coordinator) {

    }

    public Bundle toBundle() {
        Bundle bundle = new Bundle();
        bundle.putString("title", title);
        bundle.putString("description", description);
        return bundle;
    }

    @Override
    public void fromBundle(@Nullable Bundle bundle) {
        if(bundle != null) {
            title = bundle.getString("title");
            description = bundle.getString("description");
        }
    }

    public void saveTask() {
        if(!Strings.isNullOrEmpty(title) && !Strings.isNullOrEmpty(description)) {
            taskRepository.insertTask(task == null ? Task.createNewActiveTask(title, description) : task.toBuilder()
                    .setTitle(title)
                    .setDescription(description)
                    .build());
        }
    }

    public void navigateBack() {
        AddOrEditTaskKey addOrEditTaskKey = getFragment().getKey();
        if(addOrEditTaskKey.parent() instanceof TasksKey) {
            messageQueue.pushMessageTo(addOrEditTaskKey.parent(), new TasksFragment.SavedSuccessfullyMessage());
            backstack.goBack();
        } else {
            backstack.setHistory(HistoryBuilder.from(backstack.getHistory()).removeUntil(TasksKey.create()).build(), StateChange.BACKWARD);
        }
    }
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/features/addoredittask/AddOrEditTaskPresenter.kt
package com.zhuinden.simplestackdemoexamplefragments.features.addoredittask

import android.annotation.SuppressLint
import com.zhuinden.simplestack.Bundleable
import com.zhuinden.simplestack.KeyChange
import com.zhuinden.simplestackdemoexamplefragments.core.mvp.BasePresenter
import com.zhuinden.simplestackdemoexamplefragments.data.repository.TaskRepository
import com.zhuinden.simplestackdemoexamplefragments.domain.Task
import com.zhuinden.simplestackdemoexamplefragments.features.tasks.TasksFragment
import com.zhuinden.simplestackdemoexamplefragments.features.tasks.TasksKey
import com.zhuinden.simplestackdemoexamplefragments.util.BackstackHolder
import com.zhuinden.simplestackdemoexamplefragments.util.MessageQueue
import com.zhuinden.statebundle.StateBundle
import io.reactivex.android.schedulers.AndroidSchedulers
import javax.inject.Inject

/**
 * Created by Zhuinden on 2018. 08. 20.
 */
// UNSCOPED
class AddOrEditTaskPresenter @Inject constructor(
    private val taskRepository: TaskRepository,
    private val messageQueue: MessageQueue,
    private val backstackHolder: BackstackHolder
) : BasePresenter<AddOrEditTaskFragment>(), AddOrEditTaskFragment.Presenter, Bundleable {
    var title: String? = null
    var description: String? = null

    var taskId: String? = null

    var task: Task? = null

    fun updateTitle(title: String) {
        this.title = title
    }

    fun updateDescription(description: String) {
        this.description = description
    }

    @SuppressLint("CheckResult")
    override fun onAttach(view: AddOrEditTaskFragment) {
        val addOrEditTaskKey: AddOrEditTaskKey = view.getKey()
        taskId = addOrEditTaskKey.taskId

        if (!taskId.isNullOrEmpty()) {
            taskRepository.findTask(addOrEditTaskKey.taskId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { taskOptional ->
                    val task = taskOptional.takeIf { it.isPresent }?.get() ?: return@subscribe
                    this.task = task
                    if (this.title == null || this.description == null) {
                        this.title = task.title
                        this.description = task.description
                        view.setTitle(title!!)
                        view.setDescription(description!!)
                    }
                }
        }
    }

    override fun onDetach(view: AddOrEditTaskFragment) {
    }

    override fun onTitleChanged(title: String) {
        updateTitle(title)
    }

    override fun onDescriptionChanged(description: String) {
        updateDescription(description)
    }

    override fun onSaveButtonClicked() {
        if (saveTask()) {
            navigateBack()
        }
    }

    override fun toBundle(): StateBundle = StateBundle().apply {
        putString("title", title)
        putString("description", description)
    }

    override fun fromBundle(bundle: StateBundle?) {
        bundle?.run {
            title = getString("title")
            description = getString("description")
        }
    }

    private fun saveTask(): Boolean {
        val task = task
        val title = title
        val description = description

        if (title != null && title.isNotEmpty() && description != null && description.isNotEmpty()) {
            taskRepository.insertTask(when {
                taskId.isNullOrEmpty() -> Task.createNewActiveTask(title, description)
                else -> {
                    task?.copy(title = title, description = description) ?: return false
                }
            })
            return true
        }
        return false
    }

    private fun navigateBack() {
        view!!.hideKeyboard()
        val backstack = backstackHolder.backstack
        val addOrEditTaskKey = view!!.getKey<AddOrEditTaskKey>()
        when(addOrEditTaskKey) {
            is AddOrEditTaskKey.AddTaskKey -> {
                messageQueue.pushMessageTo(backstack.root<TasksKey>(), TasksFragment.SavedSuccessfullyMessage())
                backstack.goBack()
            }
            is AddOrEditTaskKey.EditTaskKey -> {
                backstack.jumpToRoot(KeyChange.BACKWARD)
            }
        }
    }
}
